### PR TITLE
[hail/ptypes] tabletype de-restrict

### PR DESCRIPTION
**Pull request by akotlar**
jvm tests don't actually trigger the condition, so need at test

### CODE REVIEW

**Comment by tpoterba** [comment]
also needs conversion code to operate on the RVDs

**Comment by akotlar** [comment]
yep looking at this now, new to it so still getting oriented

**Comment by akotlar** [comment]
interesting, no tests fail as is

**Comment by tpoterba** [comment]
need some new tests then! Let's add two tables to the src/test/resources dir, one of which has some schema that is all optional, one of which is all required. We can union them for segfaults!

**Comment by akotlar** [comment]
Hm:

Screen Shot 2020-01-31 at 10 17 03 AM

**Comment by akotlar** [comment]
Ah there is some upcast that happens before we get to case class TableUnion:

```scala
case class TableUnion(children: IndexedSeq[TableIR]) extends TableIR {
  assert(children.forall(c => c.typ.rowType == children.head.typ.rowType))
  println("ALL SAME")
```

If this is added, "ALL SAME" gets printed 16 times after "Result". Not sure why that isn't 8 or 24, but that's less important.

**Comment by tpoterba** [comment]
This should be sufficient to cause a segfault:

```
mt1 = import_vcf(file, array_elements_required=False)
mt2 = import_vcf(file, array_elements_required=False)
mt1.entries().join(mt2.entries().explode('PL').count()
```

**Comment by akotlar** [comment]
Does not work (passes) with PL present in all rows. Also does not work if I edit the VCF file and insert a bogus PL of .,.,. for each sample


Screenshot 2020-01-31 12 40 13

code in updated, pushed test

**Comment by akotlar** [comment]
Ok, have the right branch being triggered (types not the same), but causes no issues.

**Comment by tpoterba** [comment]
you also don't need to add another resource VCF to make this fail.

**Comment by akotlar** [comment]
> you also don't need to add another resource VCF to make this fail.

Yeah, I wasn't sure if the array_elements_required would be overridden if no missingness was found in the data. I wanted to test that possibility (updated my comment with the results of that test).

**Comment by akotlar** [comment]
Hm, still passes

**Comment by tpoterba** [comment]
actually, here's the test I think I want -- it probably won't segfault as written, but will write bad data:

```python
mt1 = hl.import_vcf(resource('sample.vcf'), array_elements_required=True)
mt2 = hl.import_vcf(resource('sample_missing_pl2.vcf'), array_elements_required=False)

mt3 = mt1.entries().union(mt2.entries())
mt4 = mt1.entries().union(mt2.entries())
assert mt3._same(mt3)
```

**Comment by akotlar** [comment]
Will do

**Comment by akotlar** [comment]
Still not seeing an issue with the latest version of the test:

Screenshot 2020-02-04 11 07 51

**Comment by akotlar** [comment]
this is no failing, due to subsetTO assertion

**Comment by tpoterba** [comment]
looking at this now, sorry for the delay.

**Comment by akotlar** [comment]
No worries!

**Comment by akotlar** [comment]
Yeah, so if I set deepOptional in subsetTo in the case that `virtualType != t` but `isOfType` test passes

File: hail/src/main/scala/is/hail/expr/types/physical/PType.scala [hunk 1]
```diff
@@ -295,9 +295,15 @@ abstract class PType extends Serializable with Requiredness {
       case PCanonicalInterval(p, r) =>
         val ti = t.asInstanceOf[TInterval]
         PCanonicalInterval(p.subsetTo(ti.pointType), r)
-      case _ =>
-        assert(virtualType == t)
-        this
+      case _ => {
+        if(t != virtualType && virtualType.isOfType(t)) {
+          PType.canonical(virtualType.deepOptional())
```
Comment (akotlar) [inline hail/src/main/scala/is/hail/expr/types/physical/PType.scala]: Probably risky and not what we want, but as a wip test, this does allow the table_union_mixed_requiredeness test to pass

**Comment by tpoterba** [comment]
I can't figure out why you're not hitting this:
https://github.com/hail-is/hail/blob/0399ef07e1f486c082a06b2c553702cc772670d4/hail/src/main/scala/is/hail/rvd/RVD.scala#L1423

File: hail/src/main/scala/is/hail/expr/types/physical/PType.scala (hunk 1 above)
Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/types/physical/PType.scala]: we cannot do this, no

Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/types/physical/PType.scala]: what error do you hit?

**Comment by akotlar** [comment]
Hm. Not because of the deepOptional typ setter in TableUnion?

**Comment by tpoterba** [comment]
ahh, I see. The pruner rule for union needs to change. We need a `virtualType.subsetToSameRequiredness` or something that can take the type on each side, and get its requested type from the deep-optional union type

**Comment by tpoterba** [comment]
You could probably also PR a fix to `RVD.union` that does correct type/value unification, without adjusting the type rules for TableUnion. Then you have less headache today.

**Comment by akotlar** [comment]
Got it, will push on it now

**Comment by akotlar** [comment]
addressed here: https://github.com/hail-is/hail/pull/8132

**Pull request closed without merging**